assembly/src/tokens/lines.rs

```diff
@@ -0,0 +1,502 @@
+}
+
+impl<'a> From<&'a str> for LinesStream<'a> {
+    fn from(contents: &'a str) -> Self {
+        Self {
+            lines: contents.lines(),
+            current_line: None,
+            current_line_num: 0,
+            line_char_offset: 0,
+        }
+    }
+}
+
+impl<'a> LinesStream<'a> {
+    // HELPERS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns true if the current line is a token.
+    fn is_token(&self) -> bool {
+        self.current_line
+            .filter(|line| !line.is_empty() && !line.starts_with(Token::COMMENT_PREFIX))
+            .is_some()
+    }
+
+    /// Returns true if the current line is a token or a doc comment.
+    fn is_token_or_doc_comment(&self) -> bool {
+        self.current_line
+            .filter(|line| {
+                !line.is_empty() && !line.starts_with(Token::COMMENT_PREFIX)
+                    || line.starts_with(Token::DOC_COMMENT_PREFIX)
+            })
+            .is_some()
+    }
+
+    /// Move the pointer to the next line, updating the control variables
+    fn go_to_next_line(&mut self) {
+        self.current_line = self.lines.next();
+        if let Some(line) = self.current_line {
+            let init_len = line.len();
+            let trimmed = line.trim_start();
+
+            self.current_line.replace(trimmed);
+            self.line_char_offset = (init_len - trimmed.len()) as u32;
+            self.current_line_num += 1;
+        }
+    }
+
+    /// If the current line is a doc comment, take lines until EOF or not doc comment.
+    fn take_docs_block(&mut self) -> Vec<&'a str> {
+        iter::from_fn(|| {
+            self.current_line
+                .and_then(|line| line.strip_prefix(Token::DOC_COMMENT_PREFIX))
+                .map(|doc| doc.trim())
+                .map(|doc| {
+                    self.go_to_next_line();
+                    doc
+                })
+        })
+        .fold(Vec::with_capacity(10), |mut v, doc| {
+            if !doc.trim().is_empty() {
+                v.push(doc)
+            }
+            v
+        })
+    }
+}
+
+impl<'a> Iterator for LinesStream<'a> {
+    type Item = LineInfo<'a>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.go_to_next_line();
+
+        // read next line and halt if empty
+        while !self.is_token_or_doc_comment() {
+            self.go_to_next_line();
+            self.current_line?;
+        }
+
+        // fetch a docs block, returning if not followed by a token
+        let docs = self.take_docs_block();
+        if !docs.is_empty() && !self.is_token() {
+            let line = if self.current_line.is_none() {
+                self.current_line_num
+            } else {
+                self.current_line_num.saturating_sub(1)
+            };
+            let char_offset = 0;
+            return Some(LineInfo::new(line, char_offset).with_docs(docs));
+        }
+
+        // seek next token
```

Comment:
minor nit: I'd update this comment to: "read lines until line with tokens is found; halt if empty"

Reply:
Fixed